src/models/Ruleserver.ts

```diff
@@ -0,0 +1,290 @@
+    readonly checks: Checks
+}
+
+/**
+ * The RuleServer is an experimental server that is used to propogate the rules of the watched policy rooms (BanLists) to
+ * homeservers (or e.g. synapse modules).
+ * This is done using an experimental format that is heavily based on the "Spam Checker Callbacks" made available to
+ * synapse modules https://matrix-org.github.io/synapse/latest/modules/spam_checker_callbacks.html.
+ *
+ */
+export default class Ruleserver {
+    // Each token is an index for a row of this two dimensional array. Each row represents the rules that were added during the lifetime of that token.
+    private rulesByToken: EventRules[][] = [[]];
+    private stoppedRules: string[][] = [[]];
+    // We use this to quickly lookup if we have stored a policy without scanning the rulesByToken array.
+    // First key is the room id and the second is the event id.
+    private rulesByEvent: Map<string, Map<string, EventRules>> = new Map();
+    // A unique identifier for this server instance that is given to each response so we can tell if the token
+    // was issued by this server or not. This is important for when Mjolnir has been restarted
+    // but the client consuming the rules hasn't been and we need to tell the client we have rebuilt all of the rules (via `reset` in the response).
+    private serverId: string = crypto.randomUUID();
+    private currentToken = 0;
+    private readonly banListUpdateListener = this.update.bind(this);
+    /**
+     * The token is used to seperate EventRules from each other based on when they were added.
+     * The lower the token, the longer a rule has been tracked for (relative to other rules in this Ruleserver).
+     * The token is incremented before adding new rules to be served.
+     */
+    private nextToken(): void {
+        this.currentToken += 1;
+        this.stoppedRules.push([]);
+        this.rulesByToken.push([]);
+    }
+
+    /**
+     * Get a combination of the serverId and currentToken to give to the client.
+     */
+    private get since(): string {
+        return `${this.serverId}::${this.currentToken}`;
+    }
+
+    /**
+     * Get the `EventRules` object for this Matrix event.
+     * @param roomId The room the event came from.
+     * @param eventId The id of the event.
+     * @returns The `EventRules` object describing which rules have been created based on the policy the event represents.
+     */
+    private getEventRules(roomId: string, eventId: string): EventRules|undefined {
+        return this.rulesByEvent.get(roomId)?.get(eventId);
+    }
+
+    /**
+     * Add the EventRule to be served by the ruleserver at the current token.
+     * @param eventRules Add rules for an associated policy room event. (e.g. m.policy.rule.user).
+     */
+    private addEventRules(eventRules: EventRules): void {
+        const [roomId, eventId] = [eventRules.roomId, eventRules.eventId];
+        if (this.rulesByEvent.get(roomId)?.has(eventId)) {
```

Comment:
Is there a risk that this could happen during a slow initial sync which would cause lots of updates to the `BanList`? I don't think so, but asking just in case :)

Reply:
Our BanLists don't get updated via sync at the moment (although I think I have made a point before that they probably should be) and even if they were, the mechanism we have for detecting changes within them is resilient to getting the same event more than once. This guard is really here "just in case" something changes or if we do something else with the Ruleserver in the future